### DISCUSSION TITLE
Validation: Custom attributes names of DataCollections not working

### DISCUSSION THREAD

**Discussion by wijzijnweb** [spatie/laravel-data#369] [General]
Consider this data class:

```
class RelationData extends Data
{
    public function __construct(
        public string      $name,

        #[DataCollectionOf(AddressData::class)]
        public DataCollection $addresses
    )
    {
    }

    public static function attributes(...$args): array
    {
        return [
            'name' => 'naam'
        ];
    }
}
```

It has a DataCollection of AddressData:

```
class AddressData extends Data
{
    public ?int $id;

    public ?string $code;

    public ?string $name;

    public string $street;

    public string $number;

    public string $city;

    #[DataCollectionOf(ContactData::class)]
    public DataCollection $contacts;

    public static function attributes(...$args): array
    {
        return [
            'code' => 'code',
            'street' => 'straat',
            'number' => 'huisnummer',
            'zipcode' => 'postcode',
            'city' => 'plaats'
        ];
    }
}
```

The validation error returned for address.2.city is: `Het addresses.2.city veld is verplicht.` and not `Het plaats veld is verplicht.`
Also Address has a DataCollection of ContactData that also doens't work.

What I found was that ContactData attributes where not even added. Because of this line:
https://github.com/spatie/laravel-data/blob/5b5b6590d341ab1eea35613f6deabd455d550be6/src/Resolvers/DataValidationMessagesAndAttributesResolver.php#L37

Adding: `!$dataProperty->type->isDataCollectable` makes sure the attributes are added.
But they are still not resolved.

I stranded to find out where the resolving took place.

Anyone that had the same issue?

**Comment by rubenvanassche** [comment]
Could you please provide me a pest test like this:

```php
it('tt', function (){
    class RelationData extends Data
    {
        public function __construct(
            public string      $name,

            #[DataCollectionOf(AddressData::class)]
            public DataCollection $addresses
        )
        {
        }

        public static function attributes(...$args): array
        {
            return [
                'name' => 'naam'
            ];
        }
    }

    class AddressData extends Data
    {
        public ?int $id;

        public ?string $code;

        public ?string $name;

        public string $street;

        public string $number;

        public string $city;

        #[DataCollectionOf(ContactData::class)]
        public DataCollection $contacts;

        public static function attributes(...$args): array
        {
            return [
                'code' => 'code',
                'street' => 'straat',
                'number' => 'huisnummer',
                'zipcode' => 'postcode',
                'city' => 'plaats'
            ];
        }
    }
});
```

With the whole example, including `ContactData` and a call creating the object? Assertions are not required, I just need a working example to debug further where you show me what's going on in code.

**Comment by wijzijnweb** [comment]
Hi @rubenvanassche,

This is the first Pest test I wrote. Might not be what you expected but it does the test it seems.
I'm testing if the validation error contains the custom attribute `straat`.

```
<?php

use Illuminate\Validation\ValidationException;
use Spatie\LaravelData\Attributes\DataCollectionOf;
use Spatie\LaravelData\Data;
use Spatie\LaravelData\DataCollection;
use function PHPUnit\Framework\assertStringContainsString;

it('has address validation error with custom attribute', function () {
    class RelationData extends Data
    {
        public function __construct(
            public string      $name,

            #[DataCollectionOf(AddressData::class)]
            public DataCollection $addresses
        )
        {
        }

        public static function attributes(...$args): array
        {
            return [
                'name' => 'naam'
            ];
        }
    }

    class AddressData extends Data
    {
        public ?int $id;

        public ?string $code;

        public ?string $name;

        public string $street;

        public string $number;

        public string $city;

        public static function attributes(...$args): array
        {
            return [
                'code' => 'code',
                'street' => 'straat',
                'number' => 'huisnummer',
                'zipcode' => 'postcode',
                'city' => 'plaats'
            ];
        }
    }

    $data = \json_decode('{"name":"test","addresses":[{"id":null,"street":null,"number":null,"city":null,"zipcode":null,"country":null,"default_delivery_time":"10:00","region_id":null,"is_billable":null,"is_delivery":null,"options":[],"days":[],"time_diff":[],"note":null,"order_contacts":[]}]}', true);

    try {
        RelationData::validateAndCreate($data);
    } catch (ValidationException $exception) {
        assertStringContainsString('straat', $exception->errors()['addresses.0.street'][0]);
    }
});
```

**Reply by wijzijnweb** [reply]
@rubenvanassche When you got the time, I love to hear if you can reproduce this.

**Reply by rubenvanassche** [reply]
Hi @wijzijnweb,

When I take a look at the attributes passed to the validator by laravel-data, these look like this:

<img width="355" alt="screenshot 15" src="https://user-images.githubusercontent.com/619804/227465568-17df516a-2d4c-4dbc-a16b-cc4c2709261b.png">

Which seems correct to me?

**Reply by wijzijnweb** [reply]
_Package updated to latest version 3.2.1._

@rubenvanassche Yes, I have the same and is correct I think, but the message is not correctly resolved using those attributes it seems. The actual error message that is returned is still "Het addresses.0.city veld is verplicht." instead of "Het plaats veld is verplicht."

Using the messages method does work.

```
public static function messages(...$args): array
  {

      return [
          'city.required' => 'Plaats is verplicht',
      ];

  }
  ```
  
This returns now an error of "Plaats is verplicht".


When I add the attribute in here, not using the asteriks, but the index it works as well.
So while I agree it should work with the asteriks, it doesn't but does with the index.

![image](https://user-images.githubusercontent.com/79093089/227898987-61dc0f71-6164-449c-a472-27923763b887.png)
https://github.com/spatie/laravel-data/blob/de7f02c329d2f617b0b94eaeac7e8da6c6ac869f/src/Resolvers/DataValidatorResolver.php#L37

**Comment by wijzijnweb** [comment]
I think I found a related issue in livewire, it has to with collections.

https://github.com/livewire/livewire/issues/1556

**Reply by wijzijnweb** [reply]
Or not, seems like it's already parsed to array

**Comment by rubenvanassche** [comment]
I think you should contact the people at Laravel, now I remember they're not supporting this. Even have a disabled test for it:

https://github.com/spatie/laravel-data/blob/9465cb80fd8510293539a0ccc8b21f21342e6e98/tests/ValidationTest.php#L1823

**Reply by valentinhttr** [reply]
Do you know if this is supported by Laravel in the meantime? I can't find any information on this subject.
Do you have a workaround?

**Comment by edwinhuish** [comment]
For laravel validation, this method is not static, in this case, we can do something like that:
```php
class NestRequest extends FormRequest
{
    public function rules(): array
    {
        return [
            'addresses.*.code' => 'code',
            'addresses.*.street' => 'straat',
            'addresses.*.number' => 'huisnummer',
            'addresses.*.zipcode' => 'postcode',
            'addresses.*.city' => 'plaats',
        ];
    }

    public function attributes()
    {
        $addresses = $this->attributes->get('addresses');

        $niceNames = [];

        foreach ($addresses as $idx => $address) {
            foreach ($address as $key => $val) {
                $rowIdx = $idx + 1;
                $niceNames["addresses.{$idx}.{$key}"] = "No.{$rowIdx}, {$key}";
            }
        }

        return $niceNames;
    }
}
```

How can we do something simillar in laravel data?

**Reply by rubenvanassche** [reply]
In the rules method we provide a ValidationContext which provides this kind of useful info, this is not yet the case with attributes and messages. But I'll accept a well tested and documented PR on this for sure!
